VFB-409: Table Refactoring

alexandrutrifu-urzica opened the pull request:
## What's changed
[VFB-409](https://softwiretech.atlassian.net/jira/software/c/projects/VFB/boards/1130?assignee=712020%3A6bec2b54-13b2-4f60-bfb1-44b31b4bfe1f&selectedIssue=VFB-409) The table infrastructure has been updated to use the [Material React Table](https://www.material-react-table.com/) library.

This has enabled us to implement the originally-requested **drag-and-drop functionality** inside the `Lists Page`.

## Screenshots / Videos
| Before     | After      |
|------------|------------|
| <img width="1852" height="961" alt="image" src="https://github.com/user-attachments/assets/a8b787ca-8634-4676-b269-590e1432f1b9" />| <img width="1869" height="970" alt="image" src="https://github.com/user-attachments/assets/28b143ba-d25a-4d77-9392-163f3882da90" />|
| <img width="1881" height="836" alt="image" src="https://github.com/user-attachments/assets/c433bf1b-e19e-44c3-9a50-7a7ce13c2cde" />| <img width="1892" height="874" alt="image" src="https://github.com/user-attachments/assets/9d57ce1f-b27c-45c0-aba8-731551f320b6" />|
| <img width="1869" height="930" alt="image" src="https://github.com/user-attachments/assets/58dee8fd-71eb-47e9-9dc5-b5e6461487d4" /> | <img width="1868" height="965" alt="image" src="https://github.com/user-attachments/assets/dc959221-e829-4069-b950-63535eab953e" />|
| <img width="1878" height="953" alt="image" src="https://github.com/user-attachments/assets/f1144242-7c26-443e-8809-2c1ef19ed243" />| <img width="1864" height="616" alt="image" src="https://github.com/user-attachments/assets/0a4d53ff-6f5e-47c1-a825-bb945bfaa42c" /> |
| <img width="1850" height="715" alt="image" src="https://github.com/user-attachments/assets/c5bab88b-0106-4ce8-b318-a3e5ddd19316" />| <img width="1859" height="580" alt="image" src="https://github.com/user-attachments/assets/890275ba-32e1-4eb4-ad36-3925f29cdd1d" />|
| <img width="1880" height="697" alt="image" src="https://github.com/user-attachments/assets/41372258-308f-4d79-bc38-5c1685e38f81" />| <img width="1865" height="743" alt="image" src="https://github.com/user-attachments/assets/a420e182-dcf2-41b8-a3a2-e0504a3244f9" />|
| <img width="913" height="997" alt="image" src="https://github.com/user-attachments/assets/c567cfca-c8aa-45c0-909b-34489cb32e1d" />| <img width="862" height="1006" alt="image" src="https://github.com/user-attachments/assets/70ca8705-6018-4ded-a1a9-62c6f754fd44" />|

## Checklist
- [x] The ticket is up-to-date - Please document any deviations from the original approach if there is any.
- [ ] I have documented the testing steps for QA
- [x] I have self-reviewed this PR
- [x] Make sure you've verified it works via `npm run dev`
- [x] Make sure you've verified it works via `npm run build` and `npm run start`
- [x] Make sure you've fixed all linting problems with `npm run lint_fix`
- [x] Make sure you've tested via `npm run test`

alexandrutrifu-urzica committed:
VFB-409: added new package dependencies

alexandrutrifu-urzica committed:
VFB-409: added custom cell renderer and header-to-column formatter

alexandrutrifu-urzica committed:
VFB-409: added minimal table

alexandrutrifu-urzica committed:
VFB-409: handled manual pagination and sorting

alexandrutrifu-urzica committed:
VFB-409: switched column visibility to toggleable headers (show/hide menu) and added sorting

alexandrutrifu-urzica committed:
VFB-409: started working on table docs

alexandrutrifu-urzica committed:
VFB-409: row selection through external state

alexandrutrifu-urzica committed:
VFB-409: switch to mrt, no styling yet

alexandrutrifu-urzica committed:
VFB-409: switched clients table to mrt

alexandrutrifu-urzica committed:
VFB-409: added extra condition for visibility toggle

alexandrutrifu-urzica committed:
VFB-409: added filters and refactored return object

alexandrutrifu-urzica committed:
VFB-409: refactored server-side tables

alexandrutrifu-urzica committed:
VFB-409: added some styling and row actions

alexandrutrifu-urzica committed:
VFB-409: added column styles parameter

alexandrutrifu-urzica committed:
VFB-409: row selection update

alexandrutrifu-urzica committed:
VFB-409: switched to mrt on client side tables

alexandrutrifu-urzica committed:
VFB-409: refactored testdata and types to use mrt_rowdata (temporary)

alexandrutrifu-urzica committed:
VFB-409: added parcel column styling properties

alexandrutrifu-urzica committed:
VFB-409: refactored variable name before extra column addition

alexandrutrifu-urzica committed:
VFB-409: refactored table styling and props

alexandrutrifu-urzica committed:
VFB-409: increased header font size

alexandrutrifu-urzica committed:
VFB-409: added manualSorting and manualPagination props back

alexandrutrifu-urzica committed:
VFB-409: added onRowReorder methods

alexandrutrifu-urzica committed:
VFB-409: added dividing line options and breakpoint config

alexandrutrifu-urzica committed:
VFB-409: changed styling priority and added breakpoint config prop

alexandrutrifu-urzica committed:
VFB-409: switched to mrt and updated column styles

alexandrutrifu-urzica committed:
VFB-409: refactored old table file and imports

alexandrutrifu-urzica committed:
VFB-409: updated MRT to version 2.3.0

alexandrutrifu-urzica committed:
VFB-409: added new column group type and parcel groups

alexandrutrifu-urzica committed:
VFB-409: added referral column styles

alexandrutrifu-urzica committed:
VFB-409: added column group prop to parcels table

alexandrutrifu-urzica committed:
VFB-409: removed toggleable elements from column mapper

alexandrutrifu-urzica committed:
VFB-409: removed MRT column menu and added custom component + column group prop

alexandrutrifu-urzica committed:
VFB-409: updated component for generic column groups

alexandrutrifu-urzica committed:
VFB-409: fixed linter errors

stefanghenescu-work commented on src/app/lists/ListDataview.tsx:
```diff
@@ -163,13 +200,10 @@ const ListsDataView: React.FC<ListDataViewProps> = ({
         const row1Order = row1Item.rowOrder;
```
Can you rename this to something more descriptive as it is an auxiliary value (`row1OrderAux` for example).

stefanghenescu-work commented on src/app/reports/reportsTable/styles.ts:
```diff
@@ -1,5 +1,18 @@
+    cat: { minWidth: "1rem", maxWidth: "5rem" },
+    dog: { minWidth: "1rem", maxWidth: "5rem" },
+    catAndDog: { minWidth: "1rem", maxWidth: "5rem" },
+    totalWithPets: { minWidth: "1rem", maxWidth: "5rem" },
```
Can you extract a `defaultColumnStyle` with the specified properties at the begin of the file and reuse it for every column in order to make the code cleaner.

stefanghenescu-work commented on src/components/Tables/ColumnTogglePopup.tsx:
```diff
@@ -1,61 +1,83 @@
-    shownHeaderKeys: readonly (keyof Data)[];
-    setShownHeaderKeys: (headers: (keyof Data)[]) => void;
+interface ColumnTogglePopupProps<Data extends MRT_RowData> {
+    toggleableHeaders: readonly (keyof Data | string)[];
```
Why is this a required field now?

alexandrutrifu-urzica replied:
It felt more appropriate since displaying the column menu implies there is a set of toggleable headers it needs to contain. Not passing this parameter wouldn't really be an option.

stefanghenescu-work commented on src/components/Tables/MaterialTable.tsx:
```diff
@@ -0,0 +1,458 @@
+            ),
+        muiTablePaperProps: {
+            sx: {
+                margin: "0 !important",
```
Is there any other way to do this without `!important`?

alexandrutrifu-urzica replied:
As far as I know, there is not. MUI's table styling is quite strict and there are a lot of already imposed options and dependencies between the different page elements. Forcing the override was the only way I found it possible to apply our own properties.

stefanghenescu-work commented on src/components/Tables/MaterialTable.tsx:
```diff
@@ -0,0 +1,458 @@
+                    backgroundColor: theme.main.background[1],
+                },
+            },
+        },
```
Is it possible to extract `sx` styles into reusable objects or a wrapper outside of the `useMaterialReactTable` hook?

alexandrutrifu-urzica committed:
VFB-409: updated styling for lists table

alexandrutrifu-urzica committed:
VFB-409: refactoring and table styling updates

alexandrutrifu-urzica committed:
VFB-409: updated readme

alexandrutrifu-urzica committed:
VFB-409: removed default sort config

alexandrutrifu-urzica committed:
VFB-409: updated mrt version

alexandrutrifu-urzica committed:
VFB-409: updated table tests

alexandrutrifu-urzica committed:
VFB-409: removed isLoading state

alexandrutrifu-urzica committed:
VFB-409: disabled linter for effect hook and added extra condition for pagination/sorting

# Conflicts:
#	src/app/admin/dietaryRequirementsTable/EditDietaryRequirementsModal.tsx

alexandrutrifu-urzica committed:
VFB-409: split fetch into filters and slice

alexandrutrifu-urzica committed:
VFB-409: added extra comment explaining or-condition

stonelink commented:
Closing this PR as it has been superseded by https://github.com/LambethFoodbank/foodbankapp/pull/879
